components/nexus-cellar-hazelcast/src/main/java/org/sonatype/nexus/hazelcast/HazelcastInstanceProvider.java

```diff
@@ -0,0 +1,141 @@
+import com.hazelcast.util.ExceptionUtil;
+import org.apache.karaf.cellar.core.utils.CombinedClassLoader;
+import org.eclipse.sisu.EagerSingleton;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Expose the shared {@link HazelcastInstance}.
+ *
+ * @since 3.0
+ */
+@Named(HazelcastInstanceProvider.INSTANCE_NAME)
+@EagerSingleton
+public class HazelcastInstanceProvider
+    extends ComponentSupport
+    implements Provider<HazelcastInstance>
+{
+  public static final String INSTANCE_NAME = "nexus";
+
+  static {
+    ExceptionUtil.class.toString(); // workaround https://github.com/orientechnologies/orientdb/pull/4609
+  }
+
+  private final BundleContext bundleContext;
+
+  private final ApplicationDirectories directories;
+
+  private final LocalNodeAccess localNodeAccess;
+
+  private final CombinedClassLoader combinedClassLoader;
+
+  private HazelcastInstance instance;
+
+  private ServiceRegistration<?> registration;
+
+  @Inject
+  public HazelcastInstanceProvider(final BundleContext bundleContext,
+                                   final ApplicationDirectories directories,
+                                   final LocalNodeAccess localNodeAccess)
+  {
+    this.bundleContext = checkNotNull(bundleContext);
+    this.directories = checkNotNull(directories);
+    this.localNodeAccess = checkNotNull(localNodeAccess);
+
+    combinedClassLoader = new CombinedClassLoader();
+
+    // start tracking bundles whose packages should be distributed
+    new HazelcastBundleTracker(bundleContext, combinedClassLoader);
+  }
+
+  @Override
+  public HazelcastInstance get() {
+    instance = Hazelcast.newHazelcastInstance(getHazelcastConfig());
+    checkState(instance != null, "Missing instance named: %s", INSTANCE_NAME);
+    log.info("Instance: {}", instance);
+
+    // expose hazelcast instance as service so cellar's blueprint components can consume it
+    registration = bundleContext.registerService(HazelcastInstance.class, instance, null);
+
+    return instance;
+  }
+
+  @PreDestroy
+  public void shutdown() {
+    if (instance != null) {
+      log.info("Shutting down: {}", instance);
+      try {
+        registration.unregister();
+      }
+      catch (final IllegalStateException e) {
+        // bundle has already been stopped and the service unregistered
+      }
+      finally {
+        registration = null;
+        instance.shutdown();
+        instance = null;
+      }
+    }
+  }
+
+  private Config getHazelcastConfig() {
+    Config config;
+    File hazelcastXml = new File(new File(directories.getInstallDirectory(), "etc"), "hazelcast.xml");
+    try (InputStream in = new BufferedInputStream(new FileInputStream(hazelcastXml))) {
+      config = new XmlConfigBuilder(in).build();
+    }
+    catch (IOException e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)